src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/AppInsights/AppInsightsScope.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Microsoft.Azure.WebJobs.Host.Loggers
+{
+    internal class AppInsightsScope
+    {
+        private readonly object _state;
+
+        private AppInsightsScope(object state, AppInsightsScope parent)
+        {
+            _state = state;
+            Parent = parent;
+        }
+
+        internal AppInsightsScope Parent { get; private set; }
+
+        private static AsyncLocal<AppInsightsScope> _value = new AsyncLocal<AppInsightsScope>();
+
+        public static AppInsightsScope Current
+        {
+            get
+            {
+                return _value.Value;
+            }
+            set
+            {
+                _value.Value = value;
+            }
+        }
+
+        public static IDisposable Push(object state)
+        {
+            Current = new AppInsightsScope(state, Current);
+            return new DisposableScope();
+        }
+
+        // Builds a state dictionary of all scopes. If an inner scope
+        // contains the same key as an outer scope, it overwrites the value.
+        public static IDictionary<string, object> GetMergedStateDictionary()
```

Comment:
We would typically do it using ITelemetryInitializer. You can have one attached to your TelemetryConfiguration that will initialize ALL telemetry items without an explicit call to this method